packages/forms/src/directives/select_control_value_accessor.ts

```diff
@@ -117,6 +117,35 @@ export class SelectControlValueAccessor extends BuiltInControlValueAccessor impl
   }
 
   private _compareWith: (o1: any, o2: any) => boolean = Object.is;
+  private _injector = inject(EnvironmentInjector);
+
+  /**
+   * Lifecycle method called after the view has been checked for changes. For internal use only.
+   * @nodoc
+   */
+  ngAfterViewChecked() {
+    /*
+     * This is needed to efficiently set the select value when adding/removing options. If
+     * writeValue is instead called for every added/removed option, this results in exponentially
+     * more _compareValue calls than the number of option elements (issue #41330).
+     *
+     * Secondly, calling writeValue when rendering individual option elements instead of after they
+     * are all rendered caused an issue in Safari and IE 11 where the first option element failed
+     * to be deselected when no option matched the select ngModel. This was because Angular would
+     * set the select element's value property before appending the option's child text node to the
+     * DOM (issue #14505).
+     *
+     * Finally, this approach is necessary to avoid an issue with delayed element removal when
+     * using the animations module (in all browsers). Otherwise when a selected option is removed
+     * (so no option matches the ngModel anymore), Angular would change the select element value
+     * before actually removing the option from the DOM. Then when the option is finally removed
+     * from the DOM, the browser would change the select value to that of the first option, even
+     * though it doesn't match the ngModel (issue #18430).
+     */
+    afterNextRender(() => {
+      this.writeValue(this.value);
+    }, {injector: this._injector});
```

Comment:
We could wrap this code into the `runInInjectionContext` function to resolve the issue identified by CI:
```suggestion
    runInInjectionContext(this.injector, () => {
      afterNextRender(() => {
        this.writeValue(this.value);
      });
    });
```

Reply:
Now there's an error that the call to `inject` on line 120 (to initialize `this._injector`) must be called from an injection context. How is `this.injector` in your example supposed to be initialized?

Reply:
In my example the `this.injector` is initialized in the same way you have it in your PR.

The problem with the `inject()` in this class is a test setup issue: those failing tests are not compatible with the `inject()` function, since directive instances are created via `new` ([see here](https://github.com/angular/angular/blob/main/packages/forms/test/directives_spec.ts#L85-L89 )). To fix this, those failing tests can be wrapped into a similar function (the `TestBed.runInInjectionContext` one) to establish an injection context. Here is an example:

```
diff --git a/packages/forms/test/directives_spec.ts b/packages/forms/test/directives_spec.ts
index 1bda48e1b5..114abfd01a 100644
--- a/packages/forms/test/directives_spec.ts
+++ b/packages/forms/test/directives_spec.ts
@@ -7,7 +7,7 @@
  */
 
 import {SimpleChange} from '@angular/core';
-import {fakeAsync, flushMicrotasks, tick} from '@angular/core/testing';
+import {fakeAsync, flushMicrotasks, TestBed, tick} from '@angular/core/testing';
 import {AbstractControl, CheckboxControlValueAccessor, ControlValueAccessor, DefaultValueAccessor, FormArray, FormArrayName, FormControl, FormControlDirective, FormControlName, FormGroup, FormGroupDirective, FormGroupName, NgControl, NgForm, NgModel, NgModelGroup, SelectControlValueAccessor, SelectMultipleControlValueAccessor, ValidationErrors, Validator, Validators} from '@angular/forms';
 import {selectValueAccessor} from '@angular/forms/src/directives/shared';
 import {composeValidators} from '@angular/forms/src/validators';
@@ -69,10 +69,12 @@ class CustomValidatorDirective implements Validator {
         });
 
         it('should return select accessor when provided', () => {
-          const selectAccessor = new SelectControlValueAccessor(null!, null!);
-          expect(selectValueAccessor(dir, [
-            defaultAccessor, selectAccessor
-          ])).toEqual(selectAccessor);
+          TestBed.runInInjectionContext(() => {
+            const selectAccessor = new SelectControlValueAccessor(null!, null!);
+            expect(selectValueAccessor(dir, [
+              defaultAccessor, selectAccessor
+            ])).toEqual(selectAccessor);
+          });
         });
 
         it('should return select multiple accessor when provided', () => {
@@ -83,9 +85,12 @@ class CustomValidatorDirective implements Validator {
         });
 
         it('should throw when more than one build-in accessor is provided', () => {
-          const checkboxAccessor = new CheckboxControlValueAccessor(null!, null!);
-          const selectAccessor = new SelectControlValueAccessor(null!, null!);
-          expect(() => selectValueAccessor(dir, [checkboxAccessor, selectAccessor])).toThrowError();
+          TestBed.runInInjectionContext(() => {
+            const checkboxAccessor = new CheckboxControlValueAccessor(null!, null!);
+            const selectAccessor = new SelectControlValueAccessor(null!, null!);
+            expect(() => selectValueAccessor(dir, [checkboxAccessor, selectAccessor]))
+                .toThrowError();
+          });
         });
 
         it('should return custom accessor when provided', () => {
```